### PR TITLE
[WIP] UI: Add features to settings

### PR DESCRIPTION
**Pull request by cg2121**
TODO:
-[ ] Find a better way to restart the program.

### CODE REVIEW

**Comment by Xaymar** [comment]
As far as I understand the code at 1c35ae2, this currently does not ask the user if they want to restart, and only shows them a Message Box that OBS Studio is going to restart now. This might be considered unwanted behavior by the user and result in angry users that were still considering their changes or still working on something else in the scene collection or profile.

**Comment by cg2121** [comment]
It is only restarted when the settings windows is closed, so the user can still change other settings.

**Comment by Xaymar** [comment]
That is not what I was referring to. The problem is that it restarts, and does not ask the user if they want to restart. It only tells them "I'm going to restart now. Kthxbye", with no way for the user to prevent that.

Edit: A better message would be:

> You have changed options which requires OBS Studio to be restarted for them to be applied.
> Do you want to restart OBS Studio now?
>
> [Yes] [No]

**Comment by WizardCM** [comment]
I second @Xaymar's suggestion.

**Comment by cg2121** [comment]
I've now changed it to use a confirmation dialog asking the user if they want to restart or not.

File: UI/window-basic-main.cpp [hunk 1]
```diff
@@ -467,6 +469,25 @@ static obs_data_t *GenerateSaveData(obs_data_array_t *sceneOrder,
...
+void OBSBasic::Restart()
+{
+	QMessageBox::StandardButton button =
+				OBSMessageBox::question(this,
+						QTStr("Restart"),
+						QTStr("ProgramRestarted"));
+
+	if (button == QMessageBox::Yes) {
+		qApp->quit();
+		QProcess::startDetached(qApp->arguments()[0],
```
Comment (notr1ch) [inline UI/window-basic-main.cpp]: This seems to open up a race condition where the new process tries to start before the old one is completely done shutting down. Perhaps a new startup flag is needed to make OBS wait for any other OBS processes to exit before trying to load configs etc.

**Comment by notr1ch** [comment]
Is a full restart actually necessary? Why can't we reinitialize just the audio subsystem?

**Comment by jp9000** [comment]
Reinitializing the audio subsystem is a bit of a chore right now.  All sources with audio have mixers built in to them for the audio hierarchy, so reinitializing means that you have to reinitialize all of their audio-related data.  In addition, there are also audio filters, and things like audio meters.  So it's unfortunately not quite as trivial as it seems it would be at first glance.

**Comment by SuslikV** [comment]
It is stupid thing. There is warning label in OBS that says that you need to restart application. Why you changing things in this way?

**Comment by WizardCM** [comment]
> 
> 
> It is stupid thing. There is warning label in OBS that says that you need to restart application. Why you changing things in this way?

Improves the user experience and gets them back up and running more quickly.

**Comment by cg2121** [comment]
I have now set it to ask for restart also when the language changes.

**Comment by DDRBoxman** [comment]
Any thoughts on @notr1ch 's comment about the race condition?

**Comment by cg2121** [comment]
Well, I haven't had a problem with it at all.

**Comment by cg2121** [comment]
I have now moved to quit function to below the startDetached function, just in case.

**Comment by notr1ch** [comment]
If anything I think that would make the issue worse as we want to be sure the current OBS has closed before starting another instance. Without a startup flag indicating that OBS should wait for a previous instance or handle to wait on (or similar) I don't think this can be solved from this part of the code alone.

**Comment by cg2121** [comment]
I had only tested this on Linux. When I tested it on Windows, it would crash when restarting. I changed qApp->quit() to close() and everything works just fine.

**Comment by cg2121** [comment]
I have now refactored it a bit, so that the startDetached function is the last thing called in the OBS destructor.

**Comment by cg2121** [comment]
I've combined this to #1791 because they require the same restart function.

**Comment by cg2121** [comment]
I'm going to close this for now because I've ran into bugs when restarting the program. I will reopen when I find a better way of handling it.

**Pull request closed without merging**